src/CircleEditor/CircleEditorDocument.ts

```diff
@@ -0,0 +1,197 @@
+
+  public get uri(): vscode.Uri {
+    return this._uri;
+  }
+  public get model(): Circle.ModelT {
+    return this._model;
+  }
+  public get modelData(): Uint8Array {
+    let fbb = new flatbuffers.Builder(1024);
+    Circle.Model.finishModelBuffer(fbb, this._model.pack(fbb));
+    return fbb.asUint8Array();
+  }
+
+  static async create(uri: vscode.Uri):
+      Promise<CircleEditorDocument|PromiseLike<CircleEditorDocument>> {
+    let bytes = new Uint8Array(await vscode.workspace.fs.readFile(uri));
+    return new CircleEditorDocument(uri, bytes);
+  }
+
+  private constructor(uri: vscode.Uri, bytes: Uint8Array) {
+    super();
+    this._uri = uri;
+    this._model = this.loadModel(bytes);
+  }
+
+  // dispose
+  private readonly _onDidDispose = this._register(new vscode.EventEmitter<void>());
+  public readonly onDidDispose = this._onDidDispose.event;
+
+  // tell to webview
+  public readonly _onDidChangeContent = this._register(
+      new vscode.EventEmitter<{readonly modelData: Uint8Array;}|ResponseModel|CustomInfoMessage|
+                              ResponseModelPath|ResponseFileRequest|ResponseJson>());
+  public readonly onDidChangeContent = this._onDidChangeContent.event;
+
+  // tell to vscode
+  private readonly _onDidChangeDocument = this._register(new vscode.EventEmitter<{
+    readonly label: string,
+    undo(): void,
+    redo(): void,
+  }>());
+  public readonly onDidChangeDocument = this._onDidChangeDocument.event;
+
+  dispose(): void {
+    this._onDidDispose.fire();
+    super.dispose();
+  }
+
+  private loadModel(bytes: Uint8Array): Circle.ModelT {
+    let buf = new flatbuffers.ByteBuffer(bytes);
+    return Circle.Model.getRootAsModel(buf).unpack();
+  }
+
+  /**
+   * execute appropriate edit feature and calls notifyEdit function
+   */
+  makeEdit(message: RequestMessage) {
+    const oldModelData = this.modelData;
+    switch (message.type) {
+      case 'attribute':
+        // TODO: implement functions to edit attributes of operator in model
+        break;
+      case 'tensor':
+        // TODO: implement functions to edit tensors in model;
+        break;
+      default:
+        return;
+    }
+    const newModelData = this.modelData;
+    this.notifyEdit(oldModelData, newModelData);
+  }
+
+  /**
+   * calls sendModel function so that webviews can be aware of current model data
+   * records old and new model for undo and redo features
+   */
+  notifyEdit(oldModelData: Uint8Array, newModelData: Uint8Array) {
+    this.sendModel('0');
+
+    this._onDidChangeDocument.fire({
+      label: 'Model',
+      undo: async () => {
+        this._model = this.loadModel(oldModelData);
+        this.sendModel('0');
+      },
+      redo: async () => {
+        this._model = this.loadModel(newModelData);
+        this.sendModel('0');
+      }
+    });
+  }
+
+  /**
+   * send current model data to webviews with declared packetsize
+   */
+  sendModel(offset: string) {
```

Comment:
I think `offset` is only used as int by `parseInt`.
If so, how about changing `offset` type to `number` and check with `isInteger`?

Reply:
I agree repeated parseInt function looks not good. I will modify this part!